### PR TITLE
adapt ZipDataset to buildin zip behavior and add KeyZipDataset. 

### PR DESCRIPTION
**Pull request by JanekEbb**
No short zip allowed, input_datasets have to have same lengths.

### CODE REVIEW

File: lazy_dataset/core.py [hunk 1]
```diff
@@ -147,6 +150,98 @@ def concatenate(*datasets):
...
 
+def _zip(*datasets):
+    """
+    Create a new `Dataset` zipping all passed datasets.
+
+    Example:
+        >>> import lazy_dataset
+        >>> ds1 = lazy_dataset.new({'1': 1, '2': 2, '3': 3, '4': 4})
+        >>> ds2 = lazy_dataset.new({'1': 5, '2': 6, '3': 7, '4': 8})
+        >>> _zip(ds1, ds2)
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Can you change this to `lazy_dataset.zip`?

File: lazy_dataset/core.py [hunk 2]
```diff
@@ -147,6 +150,98 @@ def concatenate(*datasets):
...
+
+    Returns:
+        zip of all input datasets
+
+    """
+    if len(datasets) == 0:
+        raise ValueError('Need at least one dataset to concatenate!')
+    if len(datasets) == 1 and isinstance(datasets[0], (tuple, list)):
+        datasets, = datasets
+    if not all(isinstance(dataset, Dataset) for dataset in datasets):
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Can you use a list comprehension instead of a generator comprehension?

File: lazy_dataset/core.py [hunk 3]
```diff
@@ -495,11 +588,11 @@ def zip(self, *others) -> 'ZipDataset':
             >>> ds4  # doctest: +ELLIPSIS
                     DictDataset(len=2)
                     DictDataset(len=2)
-                  ZipDataset()
+                  KeyZipDataset()
```
Comment (boeddeker) [inline lazy_dataset/core.py]: Should this be `ZipDataset`?

File: tests/test_zip.py [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import lazy_dataset
+
+
+def test_key_zip():
+    ds1 = lazy_dataset.new({'1': 1, '2': 2, '3': 3, '4': 4})
+    ds2 = lazy_dataset.new({'1': 5, '2': 6, '3': 7, '4': 8})
+    ds = lazy_dataset.key_zip(ds1, ds2)
+    assert list(ds) == [(1, 5), (2, 6), (3, 7), (4, 8)]
+    ds = lazy_dataset.key_zip(ds1, ds2).prefetch(2, 2)
+    assert list(ds) == [(1, 5), (2, 6), (3, 7), (4, 8)]
```
Comment (boeddeker) [inline tests/test_zip.py]: Can you add a second test, where ds2 is shuffled?

File: lazy_dataset/core.py (hunk 3 above)
Comment (JanekEbb) [reply lazy_dataset/core.py]: in that case the return of items() would not be indexable by keys

Comment (boeddeker) [reply lazy_dataset/core.py]: Ah, sorry. The context that Github showed was confusing. Your change is correct.

File: lazy_dataset/core.py [hunk 5]
```diff
@@ -147,6 +150,98 @@ def concatenate(*datasets):
...
+    """
+    if len(datasets) == 0:
+        raise ValueError('Need at least one dataset to concatenate!')
+    if len(datasets) == 1 and isinstance(datasets[0], (tuple, list)):
+        datasets, = datasets
+    if not all([isinstance(dataset, Dataset) for dataset in datasets]):
+        raise TypeError(
+            f'All input arguments must be datasets! {Dataset} ' + ' '.join(
+                str(type(d)) for d in datasets) + '|' + ' '.join(
+                str(isinstance(d, Dataset)) for d in datasets))
```
Comment (boeddeker) [inline lazy_dataset/core.py]: `str(type(d)) for d in datasets)` will yield something like `'<generator object <genexpr> at 0x7f6ea87e7308>'`.
Could you add the `[` and `]`?

Comment (boeddeker) [reply lazy_dataset/core.py]: This code is fine, I overlooked one closing bracket

**Comment by boeddeker** [comment]
Thanks, I will merge

**Pull request merged**